zebra-consensus/src/transaction/check.rs

```diff
@@ -60,19 +60,27 @@ pub fn lock_time_has_passed(
 
 /// Checks that the transaction has inputs and outputs.
 ///
+/// # Consensus
+///
 /// For `Transaction::V4`:
-/// * At least one of `tx_in_count`, `nSpendsSapling`, and `nJoinSplit` MUST be non-zero.
-/// * At least one of `tx_out_count`, `nOutputsSapling`, and `nJoinSplit` MUST be non-zero.
+///
+/// > [Sapling onward] If effectiveVersion < 5, then at least one of
+/// > tx_in_count, nSpendsSapling, and nJoinSplit MUST be nonzero.
+///
+/// > [Sapling onward] If effectiveVersion < 5, then at least one of
+/// > tx_out_count, nOutputsSapling, and nJoinSplit MUST be nonzero.
 ///
 /// For `Transaction::V5`:
-/// * This condition must hold: `tx_in_count` > 0 or `nSpendsSapling` > 0 or
-/// (`nActionsOrchard` > 0 and `enableSpendsOrchard` = 1)
-/// * This condition must hold: `tx_out_count` > 0 or `nOutputsSapling` > 0 or
-/// (`nActionsOrchard` > 0 and `enableOutputsOrchard` = 1)
 ///
-/// This check counts both `Coinbase` and `PrevOut` transparent inputs.
+/// > [NU5 onward] If effectiveVersion >= 5 then this condition MUST hold:
+/// > tx_in_count > 0 or nSpendsSapling > 0 or (nActionsOrchard > 0 and enableSpendsOrchard = 1).
 ///
-/// https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+/// > [NU5 onward] If effectiveVersion >= 5 then this condition MUST hold:
+/// > tx_out_count > 0 or nOutputsSapling > 0 or (nActionsOrchard > 0 and enableOutputsOrchard = 1).
+///
+/// <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+///
+/// This check counts both `Coinbase` and `PrevOut` transparent inputs.
```

Comment:
I would prefer to have these rules next to the code that actually implements them, so that we can see that the rules are consistent with the code.

Can we keep these rules on `has_transparent_or_shielded_inputs`, `has_transparent_or_shielded_outputs`, and `has_enough_orchard_flags` instead?

Reply:
I think this rules are fine here, anyone can follow a function. We discussed it before.

Reply:
Thanks for reminding me, I had forgotten.